target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts

```diff
@@ -0,0 +1,374 @@
+#include "qcom-ipq4019.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
+	model = "Wallystech DR40X9";
+	compatible = "wallystech,dr40x9";
+
+	chosen {
+		bootargs-append = " ubi.mtd=ubi root=/dev/ubiblock0_1";
```

Comment:
The kernel should automount the UBI volume. Or is the bootloader overwriting the bootline?

Reply:
The bootloader is passing ubi.mtd=rootfs and different root partition so it needs to overwritten

Reply:
Is this a quirk the bootloader does when booting linux images or can this behavior turned off by changing the bootline passed by the bootloader?

If so, adjusting the installation instructions should be preferred over using the `bootargs-append` hack.

Reply:
It dynamically generates the bootargs so that is why I used the bootargs-append.

Reply:
Btw., technically the "chosen ... bootargs-append" option works for us, too. It just looks a bit weird, to have root= / rootfs= listed twice, with the former, provided from u-boot, being ignored.

And we ran over some ethernet issues with the original u-boot (from QSDK?) + OpenWrt + DTS we got from Wallsys (looks slightly different then this one) combination. The probing fails and eth0/eth1 do not appear. Running a "tftpboot" or "ping" from u-boot usually helps until power-cycling. So it seems that the original u-boot (or driver / DTS combination) was missing some reset of the switch chip.

However after flashing the new u-boot provided by Wallsys via the "source ...:script" this worked too, it seems that the new u-boot we got did that reset.

But it seems you found the appropriate reset GPIO for the MDIO, your DTS looks better than the one we got from Wallsys in that regard. I guess that fixed exactly this issue for you?

Reply:
You are over complicating this.
U-boot can read the partition table from SMEM, but that is usually completely incorrect as vendors simply use one of the reference board configs.
So they simply set the U-boot env to generate the correct bootargs and that gets passed to the kernel, so you can either modify the environment or use bootargs-append to correct them from DTS.

Yeah, its weird but Linux will always take the last arguments from cmdline so you overwrite whatever vendor did that way.

I did not really use QSDK at all, simply used it to generate a running DTS and thats it.
Yeah, without the correct PHY reset GPIO Linux can sometimes be unable to properly configure ethernet as QCA U-boot tends to do some weird stuff depending on the version used.

Wallystech always provides some messed up mainline and QSDK DTS combo, which is useless, and please don't rely on the U-boot scripts for flashing.
Those only work for QSDK images.

Reply:
We are currently in contact with Wallystech to get their u-boot sources. They didn't yet provide any sources which matches their binaries and thus wasn't yet tested on a board. But the currently available things can be found at https://github.com/ecsv/u-boot-wallystech-dr4029 (see README.build for the most likely instruction how to get it build).

This can then also be used to provide a publicly available u-boot binary which just works (tm).

Reply:
> So they simply set the U-boot env to generate the correct bootargs and that gets passed to the kernel, so you can either modify the environment or use bootargs-append to correct them from DTS.

Ah! You're right, thanks to your previous findings [here](https://forum.openwrt.org/t/bootipq-and-bootm-difference-in-uboot-was-sysupgrade-from-initramfs-on-ipq4019/12547) regarding bootipq vs. bootm I now found a way which works without the "chosen ... bootargs-append". Seems the annoying "bootargs" is only hardcoded in the uboot for the "bootipq" command. Running the following instead boots fine without needing the bootargs-append and without needing to reflash the uboot:

```
(IPQ40xx) # printenv  
baudrate=115200
bootcmd=bootipq
bootdelay=2
ethact=eth0
flash_type=0
ipaddr=192.168.1.11   
machid=8010006
serverip=192.168.1.10 
stderr=serial
stdin=serial
stdout=serial

Environment size: 190/65532 bytes
(IPQ40xx) # nand device 0 && set mtdids nand0=nand0 && set mtdparts mtdparts=nand0:0x4000000@0x0(fs),${msmparts} && ubi part fs && ubi read 0x84000000 kernel
...
(IPQ40xx) # bootm
...
root@OpenWrt:/# cat /proc/cmdline 

root@OpenWrt:/# 
```

The "(IPQ40xx) # nand device 0 && ..." seems to be what is hardcoded into uboot's "bootipq".

Reply:
And, surprise, surprise, changing the "mtdparts=nand0:0x4000000@0x0(fs)" to 0x8000000 then also works for a 128MB NAND setting in the DTS and sysupgrade.